### PR TITLE
[FX] Add Interpreter and Transformer

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50420 [FX] Add Interpreter and Transformer**

1. Codify the interpreter pattern that was used, for ex, in [shape prop](https://github.com/pytorch/pytorch/blob/c3b4b2062748ed70e689f9f9ffe670e6fa20a071/torch/fx/experimental/shape_prop.py#L7). This is a pattern users will likely find repeatedly useful so it's good to have it first-class in the API. As seen in the test cases, this makes it incredibly easy to do simple analysis and transforms like swapping out individual nodes

2. Add a small `Transformer` class that is an `Interpreter` but provides a `transform` method that yields a GraphModule from interpreting the code

image
image

TODO:

- [x] Write docstrings
- [x] Write docs
- [ ] Look for more places this can be used

Differential Revision: [D25880330](https://our.internmc.facebook.com/intern/diff/D25880330)

### CODE REVIEW

**Comment by Chillee** [comment]
I think it might be easier to implement vmap with this.

File: torch/fx/interpreter.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+        arguments passed to ``run`` and this method returns
+        next() on that iterator.
+
+        Args:
+            n (Node): The placeholder node to run
+
+        Returns:
+            Any: The argument value that was retrieved.
+        """
+        return next(self.args_iter)
```
Comment (zdevito) [inline torch/fx/interpreter.py:126]: This doesn't handle `*args` correctly.

File: torch/fx/interpreter.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+    Args:
+        module (GraphModule): The module to be executed
+    """
+    def __init__(self, module : GraphModule):
+        assert isinstance(module, GraphModule)
+        self.module = module
+        self.submodules = dict(self.module.named_modules())
+        self.env : Dict[Node, Any] = {}
+
+        self.clear_env_on_return : bool = True
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/fx/interpreter.py:63]: Maybe run should just take an initial environment, and otherwise substitute a blank one?

**Review by zdevito — COMMENTED** [review]
Looks pretty reasonable. I have a few suggestions to better decompose the components into reusable parts.

File: docs/source/fx.rst [hunk 3]
```diff
@@ -322,3 +322,9 @@ API Reference
   :members:
 
 .. autoclass:: torch.fx.Proxy
+
+.. autoclass:: torch.fx.Interpreter
+  :members:
+
+.. autoclass:: torch.fx.Transformer
```
[the remark this answers is not in the document]
Comment (jamesr66a) [reply docs/source/fx.rst:329]: I haven't heard any good suggestions for this. Is the concern that this is the same name as the `Transformer` architecture? In that case, I would argue that "Transformer" is a terrible name for that architecture and I don't think we should let them squat on an English language common noun

**Review by zdevito — APPROVED** [review]

**Pull request closed without merging**